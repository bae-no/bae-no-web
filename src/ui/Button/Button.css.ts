import { createVar, style } from "@vanilla-extract/css";
import { recipe, RecipeVariants } from "@vanilla-extract/recipes";
import { sprinkles } from "../sprinkles.css";
import { theme } from "../tokens";
import { fontStyle } from "../Typography/Typography.css";

const { colors } = theme;

const defaultColorVar = createVar();
const activeColorVar = createVar();
const disabledColorVar = createVar();

const fontColorVar = createVar();
const disableFontColorVar = createVar();

const primaryStyle = style({
  backgroundColor: defaultColorVar,
  color: fontColorVar,
  ":active": {
    backgroundColor: activeColorVar,
  },
  ":disabled": {
    backgroundColor: disabledColorVar,
    color: disableFontColorVar,
  },
  selectors: {
    "&.disabled": {
      backgroundColor: disabledColorVar,
      color: disableFontColorVar,
    },
  },
});

const outlineStyle = style({
  borderColor: defaultColorVar,
  color: defaultColorVar,
  borderWidth: "1px",
  borderStyle: "solid",
  ":active": {
    borderColor: activeColorVar,
    color: activeColorVar,
  },
  selectors: {
    "&.disabled": {
      borderColor: disabledColorVar,
      color: disabledColorVar,
    },
  },
});

export const buttonStyle = recipe({
  base: sprinkles({ borderRadius: "xs" }),
  variants: {
    variant: {
      primary: primaryStyle,
      outline: outlineStyle,
    },
    color: {
      orange: {
        vars: {
          [defaultColorVar]: colors.orange2,
          [activeColorVar]: colors.orange7,
          [disabledColorVar]: colors.orange5,
          [fontColorVar]: colors.white,
          [disableFontColorVar]: colors.white,
        },
      },
      gray: {
        vars: {
          [defaultColorVar]: colors.black9,
          [activeColorVar]: colors.black11,
          [disabledColorVar]: colors.black9,
          [fontColorVar]: colors.black1,
          [disableFontColorVar]: colors.black6,
        },
      },
      white: {
        vars: {
          [defaultColorVar]: colors.white,
          [activeColorVar]: colors.white,
          [disabledColorVar]: colors.white,
          [fontColorVar]: colors.black1,
          [disableFontColorVar]: colors.black1,
        },
      },
    },
    size: {
      l: [
        sprinkles({ py: "sm", width: "full", gap: "xs" }),
        fontStyle({ fontSize: "body1-m" }),
      ],
      m: [
        sprinkles({ py: "xs", px: "md", width: "fit", gap: "xs" }),
        fontStyle({ fontSize: "body2-m" }),
      ],
      s: [
        sprinkles({ py: "xxs", px: "xs", width: "fit", gap: "xxs" }),
        fontStyle({ fontSize: "body3-m" }),
      ],
    },
  },
  defaultVariants: { variant: "primary", color: "orange", size: "l" },
});

export type ButtonStyle = RecipeVariants<typeof buttonStyle>;
