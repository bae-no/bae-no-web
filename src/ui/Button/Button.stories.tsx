import { ComponentStory, ComponentMeta } from "@storybook/react";

import Button from "./Button";

export default {
  component: Button,
  title: "UI/Button",
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: "Default",
  variant: "default",
};

export const Outline = Template.bind({});
Outline.args = {
  children: "Outline",
  variant: "outline",
};
