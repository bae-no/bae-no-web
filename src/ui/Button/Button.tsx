import Link from "next/link";
import { Box } from "../Box";
import { Icon } from "../Icon";
import { element } from "../reset.css";
import { buttonStyle } from "./Button.css";
import { ButtonProps } from "./Button.type";

function Button({
  children,
  variant,
  leftIconName,
  type,
  onClick,
  color,
  size = "l",
  href,
  disabled,
  ...rest
}: ButtonProps) {
  const renderButton = () => (
    <Box
      as={href ? "a" : "button"}
      type={type}
      alignItems="center"
      justifyContent="center"
      flexDirection="row"
      className={[
        element.button,
        buttonStyle({ variant, color, size }),
        disabled && "disabled",
      ]}
      onClick={onClick}
      disabled={disabled}
      {...rest}
    >
      {leftIconName && (
        <Icon name={leftIconName} size={size === "l" ? "lg" : "md"} />
      )}
      {children}
    </Box>
  );

  if (href && !disabled) {
    return (
      <Link passHref href={href}>
        {renderButton()}
      </Link>
    );
  }

  return renderButton();
}

export default Button;
