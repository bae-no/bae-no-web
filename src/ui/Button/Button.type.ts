import { MouseEvent, ReactNode, AriaAttributes } from "react";
import { SvgIconKey } from "../Icon/iconMap";
import { ButtonStyle } from "./Button.css";

interface ButtonBaseProps extends AriaAttributes {
  children: ReactNode;
  leftIconName?: SvgIconKey;
  disabled?: boolean;
}

interface ButtonTypeProps extends ButtonBaseProps {
  type?: "button" | "submit" | "reset";
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  href?: never;
}

interface LinkTypeProps extends ButtonBaseProps {
  href: string;
  type?: never;
  onClick?: never;
}

export type ButtonProps = ButtonStyle & (ButtonTypeProps | LinkTypeProps);
