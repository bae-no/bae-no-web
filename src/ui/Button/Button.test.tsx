import { fireEvent, render, screen } from "@testing-library/react";

import Button from "./Button";

describe("UI Button Component", () => {
  it("should render text", () => {
    render(<Button href="/">test text</Button>);
    expect(screen.getByText("test text")).toBeInTheDocument();
  });

  it("should trigger click event", () => {
    const mockClick = jest.fn();
    render(<Button onClick={mockClick}>test text</Button>);
    fireEvent.click(screen.getByRole("button"));

    expect(mockClick).toHaveBeenCalled();
  });

  it("should render link", () => {
    render(<Button href="/">test text</Button>);

    expect(screen.getByRole("link")).toBeInTheDocument();
  });

  it("should not trigger click event on disabled", () => {
    const mockClick = jest.fn();
    render(
      <Button disabled onClick={mockClick}>
        test text
      </Button>
    );
    fireEvent.click(screen.getByText("test text"));

    expect(mockClick).not.toHaveBeenCalled();
  });
});
