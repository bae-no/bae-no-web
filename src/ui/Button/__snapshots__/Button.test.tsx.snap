// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Button Component should match snapshot 1`] = `
<div>
  <a
    class="sprinkles_alignItems_center_mobile__1vna7gb2o sprinkles_justifyContent_center_mobile__1vna7gb26 sprinkles_display_flex_mobile__1vna7gb3 sprinkles_flexDirection_row_mobile__1vna7gbr reset_base__4op0mk0 reset_button__4op0mk1 Button__hhbirr7 sprinkles_borderRadius_xs_mobile__1vna7gbjf Button_buttonStyle_variant_primary__hhbirrb Button_primaryStyle__hhbirr5 Button_buttonStyle_color_orange__hhbirrd Button_buttonStyle_size_l__hhbirrg Button__hhbirr8 sprinkles_paddingTop_sm_mobile__1vna7gb40 sprinkles_paddingBottom_sm_mobile__1vna7gb4u sprinkles_width_full_mobile__1vna7gbbc sprinkles_gap_xs_mobile__1vna7gb1f Typography_base__1rfncg10 Typography_fontStyle_fontSize_body1-m__1rfncg17"
    href="/"
  >
    test
  </a>
</div>
`;
